packages/suite/src/views/firmware/index.tsx

```diff
@@ -180,406 +41,167 @@ type Props = ReturnType<typeof mapDispatchToProps> &
+            firmware.targetRelease?.release.version.join('.');
+        return <Translation id="FIRMWARE_UPDATE_TO_VERSION" values={{ version: nextVersion }} />;
+    };
 
-    const model = device.features.major_version;
+    const getComponent = () => {
+        // edge case 1 - error
+        if (firmware.error) {
+            return {
+                Heading: <ErrorStep.Heading />,
+                Body: <ErrorStep.Body />,
+                BottomBar: <CloseButton onClick={onClose} />,
+            };
+        }
+
+        // edge case 2 - user has reconnected device that is already up to date
+        if (firmware.status !== 'done' && device?.firmware === 'valid') {
+            return {
+                Heading: <NoNewFirmware.Heading />,
+                Body: <NoNewFirmware.Body />,
+                BottomBar: <CloseButton onClick={onClose} />,
+            };
+        }
+
+        switch (firmware.status) {
+            case 'initial':
+                return {
+                    Heading: <InitialStep.Heading />,
+                    Body: <InitialStep.Body />,
+                    BottomBar: <InitialStep.BottomBar />,
+                };
+            case 'check-seed':
+                return {
+                    Heading: <CommonHeading />,
+                    Body: <CheckSeedStep.Body />,
+                    BottomBar: <CheckSeedStep.BottomBar />,
+                };
+            case 'waiting-for-bootloader':
+                return {
+                    Heading: <CommonHeading />,
+                    Body: <ReconnectInBootloaderStep.Body />,
+                    BottomBar: <ReconnectInBootloaderStep.BottomBar />,
+                };
+            case 'waiting-for-confirmation':
+            case 'installing':
+            case 'started':
+            case 'downloading':
+            case 'check-fingerprint':
+            case 'wait-for-reboot':
+            case 'unplug':
+                return {
+                    Heading: <CommonHeading />,
+                    Body: <FirmwareProgressStep.Body />,
+                    BottomBar: null,
+                };
+            case 'reconnect-in-normal':
+                return {
+                    Heading: <CommonHeading />,
+                    Body: <ReconnectInNormalStep.Body />,
+                    BottomBar: <ReconnectInNormalStep.BottomBar />,
+                };
+            case 'partially-done':
+                return {
+                    Heading: <CommonHeading />,
+                    Body: <PartiallyDoneStep.Body />,
+                    BottomBar: <ContinueButton onClick={resetReducer} />,
+                };
+            case 'done':
+                return {
+                    Heading: <DoneStep.Heading />,
+                    Body: <DoneStep.Body />,
+                    BottomBar: <CloseButton onClick={onClose} />,
+                };
+
+            default:
+                // should never get here
+                return null;
```

Comment:
If we don't ever expect it to get there, we should throw an error. Just in case it happens, we can log it.

Reply:
probably reasonable